### PR TITLE
Eric/inbox header

### PR DESCRIPTION
**Pull request by EricBAndrews**
# Checklist
- [x] I have read CONTRIBUTING.md
- [x] I have described what this PR contains
- [x] If this PR alters the UI, I have attached pictures/videos
- [x] This PR addresses one or more open issues that were assigned to me:
      - progress towards #934 

# Pull Request Information

## About this Pull Request
In preparation for moderator inbox, replaces the old inbox header with a feed header matching the Feeds tab and updates its selector to use bubble pickers that stay pinned to the top of the view when scrolling. Also performs some long-overdue refactoring on the inbox view structure.

To use `FeedHeaderView`, `FeedType` is now a protocol to which `PostFeedType` and `InboxFeedType` both conform.

Note that there are some slightly funky things like the one-case `InboxFeedType` enum that only make sense given the context that moderator inbox will be added shortly.

https://github.com/mlemgroup/mlem/assets/44140166/15d0aaf6-3c62-4c93-a652-9ce8b9462823

### CODE REVIEW

File: Mlem/Views/Tabs/Feeds/Components/FeedHeaderView.swift [hunk 1]
```diff
@@ -8,27 +8,23 @@
...
+    var iconNameFill: String { get }
+    var iconScaleFactor: CGFloat { get }
+}
+
 struct FeedHeaderView: View {
     @EnvironmentObject var appState: AppState
     
-    let feedType: FeedType
+    let feedType: any FeedType
+    let suppressDropdownIndicator: Bool
```
Comment (Sjmarf) [inline Mlem/Views/Tabs/Feeds/Components/FeedHeaderView.swift:23]: Maybe this makes more sense as `showDropdownIndicator` with a default value of `true`?

File: Mlem/Views/Tabs/Inbox/Inbox View.swift [hunk 2]
```diff
@@ -146,26 +195,17 @@ struct InboxView: View {
...
-            .hoistNavigation {
-                withAnimation {
-                    scrollProxy.scrollTo(scrollToTop)
+            } header: {
+                VStack(spacing: 0) {
+                    BubblePicker(InboxTab.allCases, selected: $curTab) { tab in
+                        Text(tab.label)
+                    }
+                    .background(Color.systemBackground.opacity(scrollToTopAppeared ? 1 : 0))
+                    .background(.thickMaterial)
```
Comment (Sjmarf) [inline Mlem/Views/Tabs/Inbox/Inbox View.swift:204]:
```suggestion
                    .background(.bar)
```

Maybe this is nicer?

File: Mlem/Views/Tabs/Inbox/Inbox View.swift [hunk 3]
```diff
@@ -146,26 +195,17 @@ struct InboxView: View {
...
-                await Task {
-                    await refresh()
-                }.value
-            }
-            .hoistNavigation {
-                withAnimation {
-                    scrollProxy.scrollTo(scrollToTop)
+            } header: {
+                VStack(spacing: 0) {
+                    BubblePicker(InboxTab.allCases, selected: $curTab) { tab in
```
Comment (Sjmarf) [inline Mlem/Views/Tabs/Inbox/Inbox View.swift:200]: The vertical padding on this view doesn't match the vertical padding we use in `CommunityFeedView`. We could make the padding a little taller to match?

File: Mlem/Views/Tabs/Inbox/Inbox View.swift [hunk 4]
```diff
@@ -80,58 +117,70 @@ struct InboxView: View {
...
-            .environment(\.scrollViewProxy, scrollProxy)
-        }
-        .onChange(of: shouldFilterRead) { newValue in
-            Task(priority: .userInitiated) {
-                await handleShouldFilterReadChange(newShouldFilterRead: newValue)
+            .hoistNavigation {
+                withAnimation {
+                    scrollProxy?.scrollTo(scrollToTop)
+                }
+                return true
```
Comment (Sjmarf) [inline Mlem/Views/Tabs/Inbox/Inbox View.swift:140]: Scroll-to-top doesn't appear to be working.

**Review by Sjmarf — CHANGES_REQUESTED** [review]
Looks good! A few nits

File: Mlem/Views/Tabs/Inbox/Inbox View.swift (hunk 4 above)
Comment (EricBAndrews) [reply Mlem/Views/Tabs/Inbox/Inbox View.swift:140]: 🤦 I had tab navigation to `search` lmao

File: Mlem/Views/Tabs/Inbox/Inbox View.swift (hunk 2 above)
Comment (EricBAndrews) [reply Mlem/Views/Tabs/Inbox/Inbox View.swift:204]: Yep, that's definitely better

File: Mlem/Views/Tabs/Feeds/Components/FeedHeaderView.swift (hunk 1 above)
Comment (EricBAndrews) [reply Mlem/Views/Tabs/Feeds/Components/FeedHeaderView.swift]: Sure, that works. I think "override default behavior" parameters defaulting to `false` makes more semantic sense, but "show element" parameters defaulting to `true` does save a few `!`s--I'm not strongly opinionated on this one.

**Comment by EricBAndrews** [comment]
I have made some minor changes to `BubblePicker` and associated views to standardize the spacing across the app and improve UX:
- `BubblePicker` now comes with a default vertical padding of 10
- `BubblePicker` can now take the optional parameter `withDividers` to add a top and/or bottom divider. This allows it to be used in a context that adds spacing (e.g., `VStack(spacing: 5)`) without needing to use nested `VStack(spacing: 0)`s to achieve normal spacing between the bubbles and the dividers
- Standard 10 padding is used for `BubblePicker` in all inline contexts (inbox, community feed, instance details, user profile)
- Made the header height exactly consistent in Inbox, Feeds, and community details view (was previously a few pixels different in each)

File: Mlem/Views/Tabs/Search/BubblePicker.swift [hunk 5]
```diff
@@ -5,67 +5,88 @@
...
+enum DividerPlacement {
+    case top, bottom
+}
 
 struct BubblePicker<Value: Identifiable & Equatable & Hashable>: View {
     @Dependency(\.hapticManager) var hapticManager
     
     @Binding var selected: Value
     let tabs: [Value]
+    let dividers: [DividerPlacement]
```
Comment (Sjmarf) [inline Mlem/Views/Tabs/Search/BubblePicker.swift:20]: This should really be a `Set`, though I now realise I'm also guilty of using arrays where I shouldn't for `UserListRow` complications 😅

File: Mlem/Views/Tabs/Search/SearchResultsView.swift
Comment (Sjmarf) [inline Mlem/Views/Tabs/Search/SearchResultsView.swift:1]: We should make the spacing at the top and bottom of the `BubblePicker` here appear equal - at the moment, the bottom padding appears to be slightly thicker than the top

Screenshot 2024-03-22 at 4 12 28 pm

**Review by Sjmarf — CHANGES_REQUESTED** [review]
Nice! I like the idea of including dividers in `BubblePicker`

File: Mlem/Views/Tabs/Search/SearchResultsView.swift
Comment (EricBAndrews) [reply Mlem/Views/Tabs/Search/SearchResultsView.swift:1]: Nice spot, fixed

**Review by Sjmarf — APPROVED** [review]
LGTM 👍

**Pull request merged**